### PR TITLE
tests: run smoke tests in parallel jobs

### PR DESCRIPTION
**Pull request by brendankenny**
🔠 🔢 🏌️  time :)

Counter proposal to #10988

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -38,19 +34,6 @@ jobs:
         python -m pip install --upgrade pip
         pip install protobuf==3.7.1
 
-    # Cache yarn deps. thx https://github.com/actions/cache/blob/master/examples.md#node---yarn
```
Comment (brendankenny) [inline .github/workflows/ci.yml:41]: looking at the runs for https://github.com/GoogleChrome/lighthouse/pull/10988/checks?check_run_id=783384854 and running a bunch of them myself, it seems like caching only saves like 5 seconds out of ~30. Doesn't seem worth it.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -60,17 +43,26 @@ jobs:
...
+      run: xvfb-run --auto-servernum yarn test-bundle
+    - name: yarn test-docs
+      run: xvfb-run --auto-servernum yarn test-docs
 
     - run: yarn test-lantern
     - run: yarn test-legacy-javascript
     - run: yarn i18n:checks
     - run: yarn dogfood-lhci
 
+    - name: Upload dist/
```
Comment (brendankenny) [inline .github/workflows/ci.yml:60]: kept from #10988

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -83,3 +75,28 @@ jobs:
...
+
+  smoke:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        invert: [false, true]
+      # e.g. if smoke 0 failes, continue with smoke 1 anyway
+      fail-fast: false
+    env:
+      # The smokehouse tests run by job smoke 0. smoke-1 will run the rest.
```
Comment (paulirish) [inline .github/workflows/ci.yml:87]: whatcha thinking  with ToTChrome in the mix too?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -83,3 +75,28 @@ jobs:
 
     # Fail if any changes were written to source files (ex, from: build/build-cdt-lib.js).
     - run: git diff --exit-code
+
+  smoke:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        invert: [false, true]
```
Comment (paulirish) [inline .github/workflows/ci.yml:83]: out here i'd name this `smokeTestInvert` or something

File: lighthouse-cli/test/smokehouse/frontends/smokehouse-bin.js [hunk 5]
```diff
@@ -34,18 +34,19 @@ const runnerPaths = {
  * Determine batches of smoketests to run, based on the `requestedIds`.
  * @param {Array<Smokehouse.TestDfn>} allTestDefns
  * @param {Array<string>} requestedIds
+ * @param {{invert: boolean}} options
```
Comment (paulirish) [inline lighthouse-cli/test/smokehouse/frontends/smokehouse-bin.js:37]: grep's `-v` option is called `--invert-match` so... maybe that instead.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (brendankenny) [inline .github/workflows/ci.yml:83]: also amusing would be
```yml
matrix:
  smoketest: [a11y, errors, oopif, pwa, pwa2, pwa3, dbw, redirects, seo, offline, byte, perf, lantern, metrics, legacy-javascript, source-maps]
# ...
- name: Run smoke test
  run yarn smoke -j=1 --retries=2 ${{ matrix.smoketest }}
```
and launch all the worker jobs :)

But I'm pretty sure we [compete with the rest of the org for workers](https://help.github.com/en/actions/getting-started-with-github-actions/about-github-actions#usage-limits)?

File: .github/workflows/ci.yml (hunk 3 above)
Comment (brendankenny) [reply .github/workflows/ci.yml:87]: > whatcha thinking with ToTChrome in the mix too?

```yml
matrix:
  smokeTestInvert: [false, true]
  chrome: [stable, tot]
```
and then conditionally set `CHROME_PATH`? conditionals are awkward but that doesn't seem too bad

Comment (connorjclark) [reply .github/workflows/ci.yml:102]: I broke up the ToT and parallel jobs for a reason. Things are getting complex :)

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -38,39 +52,35 @@ jobs:
...
+      run: xvfb-run --auto-servernum yarn unit
+    - name: yarn test-clients
+      run: xvfb-run --auto-servernum yarn test-clients
+    - name: yarn test-bundle
+      run: xvfb-run --auto-servernum yarn test-bundle
+    - name: yarn test-docs
+      run: xvfb-run --auto-servernum yarn test-docs
 
     - run: yarn test-lantern
     - run: yarn test-legacy-javascript
```
Comment (connorjclark) [inline .github/workflows/ci.yml:72]: Some of these tests _can_ take a long time. That's why I split up the "special" tests (legacy js, lantern, etc.) from the core unit test and the build test–with this PR all of these are still in the same job.

**Comment by connorjclark** [comment]
How long does it take?

**Comment by paulirish** [comment]
> How long does it take?

looking at https://github.com/GoogleChrome/lighthouse/actions?query=branch%3Aparsmoke (i think the time value is the end-to-end time?)

looks like it was 9min and 10-20s.. and after splitting out basics.. it's a minute faster at 8m11s.

**Comment by brendankenny** [comment]
> looks like it was 9min and 10-20s.. and after splitting out basics.. it's a minute faster at 8m11s.

It's end to end, but the parallel jobs don't always start at once. The slowest job was 7m 55s and we can improve that :)

**Comment by paulirish** [comment]
> It's end to end,

aye. just confirmed that, too.

```js
const data = await fetch('https://api.github.com/repos/GoogleChrome/lighthouse/actions/workflows/1093967/runs?conclusion=success&status=completed&per_page=100').then(r => r.json());
data
  .workflow_runs
  .filter(e => e.conclusion === "success")
  .map(e => ((new Date(e.updated_at) - new Date(e.created_at)) / 1000 / 60).toLocaleString(undefined, {style: 'unit', unit: 'minute', unitDisplay: 'long'}))
```

image

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -81,5 +91,27 @@ jobs:
         # https://buildtracker.dev/docs/guides/github-actions#configuration
         BT_API_AUTH_TOKEN: ${{ secrets.BT_API_AUTH_TOKEN }}
 
-    # Fail if any changes were written to source files (ex, from: build/build-cdt-lib.js).
-    - run: git diff --exit-code
+  smoke:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        smokeTestInvert: [false, true]
```
Comment (connorjclark) [inline .github/workflows/ci.yml:98]: yaml keys are normally skewer-case.

**Pull request closed without merging**